// @materia/mailjet/client/esm5/components/mailjet-statistic/mailjet-statistic.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var MailjetStatisticComponent = /** @class */ (function () {
    function MailjetStatisticComponent() {
        this.counts = {
            sent: 0,
            opened: 0,
            clicked: 0,
            spam: 0,
            bounced: 0,
            blocked: 0
        };
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    MailjetStatisticComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var _this = this;
        if (changes && this.data) {
            /** @type {?} */
            var data = this.data;
            data.forEach((/**
             * @param {?} d
             * @return {?}
             */
            function (d) {
                if (d.data && d.data.length) {
                    /** @type {?} */
                    var count_1 = _this.counts[d.name.toLowerCase()] = 0;
                    d.data.forEach((/**
                     * @param {?} value
                     * @return {?}
                     */
                    function (value) {
                        count_1 = count_1 + value;
                    }));
                    _this.counts[d.name.toLowerCase()] = count_1;
                }
            }));
            this.legend = this.data.map((/**
             * @param {?} d
             * @return {?}
             */
            function (d) { return d.name; }));
            this.chartOptions = Object.assign({}, {
                legend: {
                    top: 'top',
                    left: 'center',
                    data: this.legend
                },
                grid: {
                    left: '3%',
                    right: '3%',
                    bottom: '1%',
                    containLabel: true
                },
                tooltip: {
                    trigger: 'axis'
                },
                xAxis: {
                    boundaryGap: false,
                    type: 'category',
                    data: this.xAbsis
                },
                yAxis: {
                    type: 'value'
                },
                series: this.data
            });
        }
    };
    MailjetStatisticComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mailjet-statistic',
                    template: "<div fxFlex fxLayout=\"row\" *ngIf=\"! loading && data && data.length\">\r\n  <div fxLayout=\"column\" fxLayoutAlign=\"start center\" style=\"padding: 5px 24px; overflow: auto;\">\r\n    <mailjet-stats-counter backgroundColor=\"#2196F3\" width=\"80\" height=\"60\" color=\"white\" label=\"Sent\" [count]=\"counts['sent']\"></mailjet-stats-counter>\r\n    <mailjet-stats-counter backgroundColor=\"#4CAF50\" width=\"80\" height=\"60\" color=\"white\" label=\"Opened\" [count]=\"counts['opened']\"></mailjet-stats-counter>\r\n    <mailjet-stats-counter backgroundColor=\"#8BC34A\" width=\"80\" height=\"60\" color=\"white\" label=\"Clicked\" [count]=\"counts['clicked']\"></mailjet-stats-counter>\r\n    <mailjet-stats-counter backgroundColor=\"#FFC107\" width=\"80\" height=\"60\" color=\"white\" label=\"Spam\" [count]=\"counts['spam']\"></mailjet-stats-counter>\r\n    <mailjet-stats-counter backgroundColor=\"#F44336\" width=\"80\" height=\"60\" color=\"white\" label=\"Bounce\" [count]=\"counts['bounce']\"></mailjet-stats-counter>\r\n    <mailjet-stats-counter backgroundColor=\"black\" width=\"80\" height=\"60\" color=\"white\" label=\"Blocked\" [count]=\"counts['blocked']\"></mailjet-stats-counter>\r\n  </div>\r\n  <div fxFlex fxLayout=\"column\" class=\"chart-section\">\r\n    <div class=\"chart-container\">\r\n      <div *ngIf=\"xAbsis && data\" style=\"height: 100%; width: 100%;\" echarts\r\n        [options]=\"chartOptions\">\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div>\r\n<div fxFlex fxLayout=\"row\" fxLayoutAlign=\"center center\" *ngIf=\"loading\">\r\n  <mat-progress-spinner mode=\"indeterminate\"></mat-progress-spinner>\r\n</div>\r\n<div fxFlex fxLayout=\"row\" fxLayoutAlign=\"center center\" *ngIf=\" ! loading && data && ! data.length\">\r\n  <p>No statistics for this period</p>\r\n</div>\r\n",
                    styles: [":host{display:flex;width:100%;flex:1}.chart-section{position:relative}.chart-container{position:absolute;top:0;bottom:15px;left:24px;right:24px;display:flex}"]
                }] }
    ];
    /** @nocollapse */
    MailjetStatisticComponent.ctorParameters = function () { return []; };
    MailjetStatisticComponent.propDecorators = {
        data: [{ type: Input }],
        loading: [{ type: Input }],
        expanded: [{ type: Input }],
        xAbsis: [{ type: Input }]
    };
    return MailjetStatisticComponent;
}());
export { MailjetStatisticComponent };
if (false) {
    /** @type {?} */
    MailjetStatisticComponent.prototype.data;
    /** @type {?} */
    MailjetStatisticComponent.prototype.loading;
    /** @type {?} */
    MailjetStatisticComponent.prototype.expanded;
    /** @type {?} */
    MailjetStatisticComponent.prototype.xAbsis;
    /** @type {?} */
    MailjetStatisticComponent.prototype.counts;
    /** @type {?} */
    MailjetStatisticComponent.prototype.legend;
    /** @type {?} */
    MailjetStatisticComponent.prototype.chartOptions;
}
