// @materia/mailjet/client/esm5/components/stats-counter/stats-counter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var StatsCounterComponent = /** @class */ (function () {
    function StatsCounterComponent() {
    }
    StatsCounterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mailjet-stats-counter',
                    template: "<div fxLayout=\"column\" fxLayoutAlign=\"center center\"\r\n  [style.background]=\"backgroundColor\"\r\n  [style.color]=\"color\"\r\n  [style.height.px]=\"height\"\r\n  [style.width.px]=\"width\"\r\n>\r\n  <div class=\"count\">{{ count || 0 }}</div>\r\n  <div class=\"label\">{{ label }}</div>\r\n</div>",
                    styles: ["div{text-align:center}.count{font-size:1.5em}.label{font-size:1em}"]
                }] }
    ];
    StatsCounterComponent.propDecorators = {
        backgroundColor: [{ type: Input }],
        color: [{ type: Input }],
        count: [{ type: Input }],
        label: [{ type: Input }],
        width: [{ type: Input }],
        height: [{ type: Input }]
    };
    return StatsCounterComponent;
}());
export { StatsCounterComponent };
if (false) {
    /** @type {?} */
    StatsCounterComponent.prototype.backgroundColor;
    /** @type {?} */
    StatsCounterComponent.prototype.color;
    /** @type {?} */
    StatsCounterComponent.prototype.count;
    /** @type {?} */
    StatsCounterComponent.prototype.label;
    /** @type {?} */
    StatsCounterComponent.prototype.width;
    /** @type {?} */
    StatsCounterComponent.prototype.height;
}
