// @materia/users/client/esm5/signup-form/signup-form.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Output, EventEmitter, ViewChild, TemplateRef, Input } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
var SignupFormComponent = /** @class */ (function () {
    function SignupFormComponent(fb) {
        this.fb = fb;
        this.signup = new EventEmitter();
        this.cancel = new EventEmitter();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    SignupFormComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.profileFields && changes.profileFields.currentValue) {
            this.initForm();
        }
    };
    /**
     * @return {?}
     */
    SignupFormComponent.prototype.initForm = /**
     * @return {?}
     */
    function () {
        this.signupForm = this.fb.group(Object.assign({}, {
            email: ['', Validators.required],
            password: ['', Validators.required]
        }, this.getProfileFormGroup()));
    };
    /**
     * @return {?}
     */
    SignupFormComponent.prototype.getProfileFormGroup = /**
     * @return {?}
     */
    function () {
        if (!this.profileFields) {
            return {};
        }
        var /** @type {?} */ result = {};
        this.profileFields.forEach(function (field) {
            if (field.required) {
                result[field.name] = ['', Validators.required];
            }
            else {
                result[field.name] = '';
            }
        });
        return result;
    };
    /**
     * @return {?}
     */
    SignupFormComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initForm();
    };
    /**
     * @return {?}
     */
    SignupFormComponent.prototype.signupClick = /**
     * @return {?}
     */
    function () {
        this.signup.emit(this.signupForm.value);
    };
    SignupFormComponent.decorators = [
        { type: Component, args: [{
                    selector: 'materia-signup-form',
                    template: "<ng-template #template>\n\t<mat-card fxLayout=\"column\">\n\t\t<mat-toolbar>Create a new user</mat-toolbar>\n\t\n\t\t<form [formGroup]=\"signupForm\" fxLayout=\"column\" style=\"padding: 24px;\">\n\t\t\t<mat-form-field>\n\t\t\t\t<input formControlName=\"email\" matInput placeholder=\"Email\" required type=\"email\" name=\"email\">\n\t\t\t</mat-form-field>\n\t\n\t\t\t<mat-form-field>\n\t\t\t\t<input formControlName=\"password\" matInput placeholder=\"Password\" required type=\"password\" name=\"password\">\n\t\t\t</mat-form-field>\n\t\t\t<div *ngIf=\"profileFields.length > 0\">\n\t\t\t\t<materia-param-input [showCheckbox]=\"false\" [form]=\"signupForm\" [param]=\"param\" *ngFor=\"let param of profileFields\"></materia-param-input>\n\t\t\t</div>\n\n\t\t\t<div>\n\t\t\t\t<button (click)=\"signup.emit(signupForm.value)\" mat-raised-button color=\"primary\">Create</button>\n\t\t\t\t<button (click)=\"cancel.emit()\" mat-raised-button>Cancel</button>\n\t\t\t</div>\n\t\t</form>\n\t</mat-card>\n</ng-template>",
                    styles: ["mat-card{width:300px;margin:0p;padding:0}"]
                },] },
    ];
    /** @nocollapse */
    SignupFormComponent.ctorParameters = function () { return [
        { type: FormBuilder, },
    ]; };
    SignupFormComponent.propDecorators = {
        "signup": [{ type: Output },],
        "cancel": [{ type: Output },],
        "template": [{ type: ViewChild, args: ['template',] },],
        "profileFields": [{ type: Input },],
    };
    return SignupFormComponent;
}());
export { SignupFormComponent };
function SignupFormComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    SignupFormComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    SignupFormComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    SignupFormComponent.propDecorators;
    /** @type {?} */
    SignupFormComponent.prototype.signup;
    /** @type {?} */
    SignupFormComponent.prototype.cancel;
    /** @type {?} */
    SignupFormComponent.prototype.template;
    /** @type {?} */
    SignupFormComponent.prototype.signupForm;
    /** @type {?} */
    SignupFormComponent.prototype.profileFields;
    /** @type {?} */
    SignupFormComponent.prototype.fb;
}
