// @materia/users/client/esm5/user-management.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { FlexLayoutModule } from '@angular/flex-layout';
import { MatButtonModule, MatRippleModule, MatSnackBarModule, MatCardModule, MatIconModule, MatDialogModule, MatInputModule, MatListModule, MatTooltipModule, MatToolbarModule, MatDatepickerModule, MatExpansionModule, MatNativeDateModule, MatRadioModule, MatStepperModule, MatSelectModule, MatCheckboxModule, MatProgressSpinnerModule } from '@angular/material';
import { Addon } from '@materia/addons';
import { UserManagementViewComponent } from './user-management/user-management.component';
import { HttpClientModule } from '@angular/common/http';
import { SignupFormComponent } from './signup-form/signup-form.component';
import { UserManagementSetupComponent } from './user-management-setup/user-management-setup.component';
import { ParamInputComponent } from './param-input/param-input.component';
import { EmailSettingsComponent } from './emails-settings/email-settings.component';
var UserManagementModule = /** @class */ (function () {
    function UserManagementModule() {
    }
    UserManagementModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        ReactiveFormsModule,
                        HttpClientModule,
                        FlexLayoutModule,
                        MatRippleModule,
                        MatButtonModule,
                        MatSnackBarModule,
                        MatCardModule,
                        MatIconModule,
                        MatDialogModule,
                        MatInputModule,
                        MatTooltipModule,
                        MatCheckboxModule,
                        MatSelectModule,
                        MatToolbarModule,
                        MatDatepickerModule,
                        MatNativeDateModule,
                        MatExpansionModule,
                        MatListModule,
                        MatRadioModule,
                        MatStepperModule,
                        MatProgressSpinnerModule
                    ],
                    declarations: [
                        UserManagementViewComponent,
                        EmailSettingsComponent,
                        UserManagementSetupComponent,
                        SignupFormComponent,
                        ParamInputComponent
                    ],
                    exports: [UserManagementViewComponent],
                    entryComponents: []
                },] },
    ];
    UserManagementModule = tslib_1.__decorate([
        Addon('@materia/users')
    ], UserManagementModule);
    return UserManagementModule;
}());
export { UserManagementModule };
function UserManagementModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    UserManagementModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    UserManagementModule.ctorParameters;
}
