// @materia/mailjet/client/esm5/dialogs/template-editor/template-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, ViewChild, TemplateRef, EventEmitter } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
var TemplateEditorComponent = /** @class */ (function () {
    function TemplateEditorComponent(fb) {
        this.fb = fb;
        this.cancelled = new EventEmitter();
        this.confirmed = new EventEmitter();
    }
    /**
     * @return {?}
     */
    TemplateEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.refreshTemplateForm();
    };
    /**
     * @return {?}
     */
    TemplateEditorComponent.prototype.refreshTemplateForm = /**
     * @return {?}
     */
    function () {
        this.templateForm = this.fb.group({
            name: ['', Validators.required],
            author: [this.user ? this.user.Username : '', Validators.required],
            templateType: ['transactional', Validators.required]
        });
    };
    /**
     * @return {?}
     */
    TemplateEditorComponent.prototype.confirm = /**
     * @return {?}
     */
    function () {
        this.confirmed.emit(this.templateForm.value);
    };
    /**
     * @return {?}
     */
    TemplateEditorComponent.prototype.cancel = /**
     * @return {?}
     */
    function () {
        this.cancelled.emit();
    };
    TemplateEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mailjet-template-editor',
                    template: "<ng-template #templateDialog>\r\n  <mat-toolbar class=\"mailjet-theme\">\r\n    <mat-icon>\r\n      chrome_reader_mode\r\n    </mat-icon>\r\n    <h1>{{ edition ? 'Edit' : 'New' }} template</h1>\r\n  </mat-toolbar>\r\n  <form fxLayout=\"column\" [formGroup]=\"templateForm\">\r\n    <mat-form-field>\r\n      <input matInput placeholder=\"Template name\" formControlName=\"name\">\r\n    </mat-form-field>\r\n    <mat-form-field>\r\n      <input matInput placeholder=\"Author\" formControlName=\"author\">\r\n    </mat-form-field>\r\n    <mat-form-field>\r\n      <mat-select placeholder=\"Template type\" formControlName=\"templateType\">\r\n        <mat-option value=\"marketing\">\r\n          Marketing\r\n        </mat-option>\r\n        <mat-option value=\"transactional\">\r\n          Transactional\r\n        </mat-option>\r\n        <mat-option value=\"automation\">\r\n          Automation\r\n        </mat-option>\r\n      </mat-select>\r\n    </mat-form-field>\r\n    <div fxLayout=\"row\" fxLayoutAlign=\"center center\">\r\n      <button mat-raised-button class=\"mailjet-theme\" (click)=\"confirm()\">Confirm</button>\r\n      <button mat-raised-button (click)=\"cancel()\">Cancel</button>\r\n    </div>\r\n  </form>\r\n</ng-template>\r\n",
                    styles: ["button{margin:5px}.mailjet-theme{background:#ffa726;color:#fff}form{padding:10px 15px}mat-icon{margin:8px}mat-form-field{width:180px}"]
                }] }
    ];
    /** @nocollapse */
    TemplateEditorComponent.ctorParameters = function () { return [
        { type: FormBuilder }
    ]; };
    TemplateEditorComponent.propDecorators = {
        edition: [{ type: Input }],
        settings: [{ type: Input }],
        user: [{ type: Input }],
        cancelled: [{ type: Output }],
        confirmed: [{ type: Output }],
        template: [{ type: ViewChild, args: ['templateDialog',] }]
    };
    return TemplateEditorComponent;
}());
export { TemplateEditorComponent };
if (false) {
    /** @type {?} */
    TemplateEditorComponent.prototype.templateForm;
    /** @type {?} */
    TemplateEditorComponent.prototype.edition;
    /** @type {?} */
    TemplateEditorComponent.prototype.settings;
    /** @type {?} */
    TemplateEditorComponent.prototype.user;
    /** @type {?} */
    TemplateEditorComponent.prototype.cancelled;
    /** @type {?} */
    TemplateEditorComponent.prototype.confirmed;
    /** @type {?} */
    TemplateEditorComponent.prototype.template;
    /**
     * @type {?}
     * @private
     */
    TemplateEditorComponent.prototype.fb;
}
