// @materia/mailjet/client/esm5/dialogs/send-modal/send-modal.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, TemplateRef, Output, Input, EventEmitter } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
var SendModalComponent = /** @class */ (function () {
    function SendModalComponent(fb) {
        this.fb = fb;
        this.type = 'simple';
        this.confirmed = new EventEmitter();
        this.cancelled = new EventEmitter();
    }
    Object.defineProperty(SendModalComponent.prototype, "templateVariablesControl", {
        get: /**
         * @return {?}
         */
        function () {
            return ((/** @type {?} */ (this.sendForm.get('variables'))));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    SendModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.refreshForm();
    };
    /**
     * @return {?}
     */
    SendModalComponent.prototype.addTemplateVariable = /**
     * @return {?}
     */
    function () {
        this.templateVariablesControl.push(this.fb.group({
            name: [null, Validators.required],
            value: [null, Validators.required]
        }));
    };
    /**
     * @param {?} index
     * @return {?}
     */
    SendModalComponent.prototype.removeTemplateVariable = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.templateVariablesControl.removeAt(index);
    };
    /**
     * @param {?=} type
     * @return {?}
     */
    SendModalComponent.prototype.refreshForm = /**
     * @param {?=} type
     * @return {?}
     */
    function (type) {
        if (type) {
            this.type = type;
        }
        if (this.type === 'simple') {
            this.sendForm = this.fb.group({
                to: [this.to ? this.to : '', Validators.required],
                subject: [this.subject ? this.subject : '', Validators.required],
                body: [''],
                variables: this.fb.array([])
            });
        }
        else {
            if (!this.templateId) {
                this.templateId = this.templates && this.templates.length ? this.templates[0].ID : '';
            }
            this.sendForm = this.fb.group({
                to: [this.to ? this.to : '', Validators.required],
                subject: [this.subject ? this.subject : '', Validators.required],
                template: [this.templateId, Validators.required],
                variables: this.fb.array([])
            });
        }
    };
    /**
     * @return {?}
     */
    SendModalComponent.prototype.send = /**
     * @return {?}
     */
    function () {
        if (this.sendForm.valid) {
            /** @type {?} */
            var data = Object.assign({}, { type: this.type }, this.sendForm.value);
            this.confirmed.emit(data);
        }
    };
    SendModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mailjet-send-modal',
                    template: "<ng-template #sendDialog>\r\n  <mat-toolbar class=\"mailjet-theme\">\r\n    <mat-icon>\r\n      send\r\n    </mat-icon>\r\n    <h1>Send an email</h1>\r\n  </mat-toolbar>\r\n  <div fxLayout=\"column\">\r\n    <form fxLayout=\"column\" [formGroup]=\"sendForm\" (submit)=\"send()\">\r\n      <div fxLayout=\"row\" fxLayoutAlign=\"center center\">\r\n        <mat-button-toggle-group #messageTypeGroup=\"matButtonToggleGroup\" [value]=\"type\" (valueChange)=\"refreshForm($event)\">\r\n          <mat-button-toggle value=\"simple\">Simple</mat-button-toggle>\r\n          <mat-button-toggle value=\"template\" [disabled]=\"! templates?.length\">Template</mat-button-toggle>\r\n        </mat-button-toggle-group>\r\n      </div>\r\n      <mat-form-field>\r\n        <input matInput placeholder=\"Subject\" formControlName=\"subject\" required>\r\n        <mat-error>This field is required.</mat-error>\r\n      </mat-form-field>\r\n      <mat-form-field *ngIf=\"type === 'simple'\">\r\n        <textarea matInput placeholder=\"Body\" formControlName=\"body\"></textarea>\r\n      </mat-form-field>\r\n      <mat-form-field *ngIf=\"type === 'template'\">\r\n        <mat-select placeholder=\"Template\" formControlName=\"template\">\r\n          <mat-option *ngFor=\"let template of templates\" [value]=\"template.ID\">\r\n            {{ template.Name }}\r\n          </mat-option>\r\n        </mat-select>\r\n      </mat-form-field>\r\n      <mat-form-field>\r\n        <input matInput placeholder=\"To\" formControlName=\"to\" required>\r\n        <mat-error>This field is required.</mat-error>\r\n      </mat-form-field>\r\n      <div fxLayout=\"column\" class=\"template-variables\">\r\n        <div fxLayout=\"row\" fxLayoutAlign=\"start center\">\r\n          <span [matBadge]=\"sendForm.controls['variables']?.value?.length || 0\" matBadgeOverlap=\"false\">Template variables</span>\r\n          <span fxFlex>\r\n          </span>\r\n          <button mat-icon-button type=\"button\" (click)=\"addTemplateVariable()\">\r\n            <mat-icon>add</mat-icon>\r\n          </button>\r\n        </div>\r\n        <div fxLayout=\"row\" *ngFor=\"let creds of sendForm.get('variables')['controls']; let i = index\" formArrayName=\"variables\">\r\n          <ng-container [formGroupName]=\"i\">\r\n            <mat-form-field>\r\n              <input matInput placeholder=\"Name\" formControlName=\"name\" required>\r\n              <mat-error>This field is required.</mat-error>\r\n            </mat-form-field>\r\n            <mat-form-field>\r\n              <input matInput placeholder=\"Value\" formControlName=\"value\" required>\r\n              <mat-error>This field is required.</mat-error>\r\n            </mat-form-field>\r\n          </ng-container>\r\n          <button mat-icon-button type=\"button\" (click)=\"removeTemplateVariable(i)\">\r\n            <mat-icon>delete</mat-icon>\r\n          </button>\r\n        </div>\r\n      </div>\r\n      <div fxLayout=\"row\" fxLayoutAlign=\"start center\" [style.marginTop.px]=\"15\">\r\n        <button mat-raised-button type=\"submit\" class=\"mailjet-theme\">Send</button>\r\n        <button mat-raised-button type=\"button\" (click)=\"cancelled.emit()\">Cancel</button>\r\n      </div>\r\n    </form>\r\n  </div>\r\n</ng-template>\r\n",
                    styles: [".mailjet-theme{background:#ffa726;color:#fff}mat-toolbar mat-icon{margin:8px}form{padding:10px 15px;width:300px}form mat-button-toggle-group{margin-bottom:15px}form mat-form-field{margin:0 5px}form .template-variables{margin-top:15px}form .template-variables mat-form-field{width:100px}form button{margin:5px}form::ng-deep .mat-badge-content{color:#fff;background:#ffa726}"]
                }] }
    ];
    /** @nocollapse */
    SendModalComponent.ctorParameters = function () { return [
        { type: FormBuilder }
    ]; };
    SendModalComponent.propDecorators = {
        to: [{ type: Input }],
        templateId: [{ type: Input }],
        templates: [{ type: Input }],
        subject: [{ type: Input }],
        type: [{ type: Input }],
        confirmed: [{ type: Output }],
        cancelled: [{ type: Output }],
        template: [{ type: ViewChild, args: ['sendDialog',] }]
    };
    return SendModalComponent;
}());
export { SendModalComponent };
if (false) {
    /** @type {?} */
    SendModalComponent.prototype.sendForm;
    /** @type {?} */
    SendModalComponent.prototype.to;
    /** @type {?} */
    SendModalComponent.prototype.templateId;
    /** @type {?} */
    SendModalComponent.prototype.templates;
    /** @type {?} */
    SendModalComponent.prototype.subject;
    /** @type {?} */
    SendModalComponent.prototype.type;
    /** @type {?} */
    SendModalComponent.prototype.confirmed;
    /** @type {?} */
    SendModalComponent.prototype.cancelled;
    /** @type {?} */
    SendModalComponent.prototype.template;
    /**
     * @type {?}
     * @private
     */
    SendModalComponent.prototype.fb;
}
