// @materia/server/lib/entities/queries/utils/condition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const error_1 = require("../../../error");
class Condition {
    constructor(condition, parentEntity) {
        if (!condition.name || !condition.operator) {
            throw new error_1.MateriaError('Missing required parameter to build a condition');
        }
        this.entity = condition.entity || parentEntity;
        this.parent = parentEntity;
        this.name = condition.name;
        this.operator = condition.operator;
        this.value = condition.value == undefined ? '' : condition.value;
        this.operand = condition.operand;
    }
    valueIsParam() {
        return (typeof this.value == 'string' && this.value.length > 0 && this.value[0] == '=');
    }
    toJson() {
        const res = {
            name: this.name,
            operator: this.operator
        };
        if (this.operator.toUpperCase() != 'IS NULL' && this.operator.toUpperCase() != 'IS NOT NULL') {
            res.value = this.value;
        }
        if (this.entity && this.parent && this.entity != this.parent) {
            res.entity = this.entity;
        }
        if (this.operand) {
            res.operand = this.operand;
        }
        return res;
    }
}
exports.Condition = Condition;
