// @materia/users/client/esm2015/signup-form/signup-form.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Output, EventEmitter, ViewChild, TemplateRef, Input } from '@angular/core';
import { FormBuilder, Validators } from '@angular/forms';
export class SignupFormComponent {
    /**
     * @param {?} fb
     */
    constructor(fb) {
        this.fb = fb;
        this.signup = new EventEmitter();
        this.cancel = new EventEmitter();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.profileFields && changes.profileFields.currentValue) {
            this.initForm();
        }
    }
    /**
     * @return {?}
     */
    initForm() {
        this.signupForm = this.fb.group(Object.assign({}, {
            email: ['', Validators.required],
            password: ['', Validators.required]
        }, this.getProfileFormGroup()));
    }
    /**
     * @return {?}
     */
    getProfileFormGroup() {
        if (!this.profileFields) {
            return {};
        }
        const /** @type {?} */ result = {};
        this.profileFields.forEach(field => {
            if (field.required) {
                result[field.name] = ['', Validators.required];
            }
            else {
                result[field.name] = '';
            }
        });
        return result;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.initForm();
    }
    /**
     * @return {?}
     */
    signupClick() {
        this.signup.emit(this.signupForm.value);
    }
}
SignupFormComponent.decorators = [
    { type: Component, args: [{
                selector: 'materia-signup-form',
                template: `<ng-template #template>
	<mat-card fxLayout="column">
		<mat-toolbar>Create a new user</mat-toolbar>
	
		<form [formGroup]="signupForm" fxLayout="column" style="padding: 24px;">
			<mat-form-field>
				<input formControlName="email" matInput placeholder="Email" required type="email" name="email">
			</mat-form-field>
	
			<mat-form-field>
				<input formControlName="password" matInput placeholder="Password" required type="password" name="password">
			</mat-form-field>
			<div *ngIf="profileFields.length > 0">
				<materia-param-input [showCheckbox]="false" [form]="signupForm" [param]="param" *ngFor="let param of profileFields"></materia-param-input>
			</div>

			<div>
				<button (click)="signup.emit(signupForm.value)" mat-raised-button color="primary">Create</button>
				<button (click)="cancel.emit()" mat-raised-button>Cancel</button>
			</div>
		</form>
	</mat-card>
</ng-template>`,
                styles: [`mat-card{width:300px;margin:0p;padding:0}`]
            },] },
];
/** @nocollapse */
SignupFormComponent.ctorParameters = () => [
    { type: FormBuilder, },
];
SignupFormComponent.propDecorators = {
    "signup": [{ type: Output },],
    "cancel": [{ type: Output },],
    "template": [{ type: ViewChild, args: ['template',] },],
    "profileFields": [{ type: Input },],
};
function SignupFormComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    SignupFormComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    SignupFormComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    SignupFormComponent.propDecorators;
    /** @type {?} */
    SignupFormComponent.prototype.signup;
    /** @type {?} */
    SignupFormComponent.prototype.cancel;
    /** @type {?} */
    SignupFormComponent.prototype.template;
    /** @type {?} */
    SignupFormComponent.prototype.signupForm;
    /** @type {?} */
    SignupFormComponent.prototype.profileFields;
    /** @type {?} */
    SignupFormComponent.prototype.fb;
}
