// @materia/users/client/esm5/emails-settings/email-settings.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { FormBuilder, Validators } from '@angular/forms';
import { MatExpansionPanel } from '@angular/material';
var EmailSettingsComponent = /** @class */ (function () {
    function EmailSettingsComponent(http, form) {
        this.http = http;
        this.form = form;
        this.hide = new EventEmitter();
        this.save = new EventEmitter();
    }
    /**
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    EmailSettingsComponent.prototype.getSettingsProperty = /**
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    function (property, defaultValue) {
        return (this.settings && this.settings[property]) || defaultValue;
    };
    /**
     * @return {?}
     */
    EmailSettingsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.emailForm = this.form.group({
            method: [
                this.getSettingsProperty('method', 'session'),
                Validators.required
            ],
            static_salt: [
                this.getSettingsProperty('static_salt', ''),
                Validators.required
            ],
            user_profile_enabled: [
                this.getSettingsProperty('user_profile_enabled', 'false')
            ],
            user_profile_entity: [
                this.getSettingsProperty('user_profile_entity', '')
            ],
            email_verification: [
                this.getSettingsProperty('email_verification', false)
            ],
            email_addon: [this.getSettingsProperty('email_addon', false)],
            template_signup: [this.getSettingsProperty('template_signup', '')],
            redirect_signup: [this.getSettingsProperty('redirect_signup', '')],
            subject_signup: [this.getSettingsProperty('subject_signup', '')],
            subject_lost_password: [this.getSettingsProperty('subject_lost_password', '')],
            subject_change_email: [this.getSettingsProperty('subject_change_email', '')],
            template_lost_password: [
                this.getSettingsProperty('template_lost_password', '')
            ],
            redirect_lost_password: [
                this.getSettingsProperty('redirect_lost_password', '')
            ],
            template_change_email: [
                this.getSettingsProperty('template_change_email', '')
            ],
            redirect_change_email: [
                this.getSettingsProperty('redirect_change_email', '')
            ]
        });
        this.refreshTemplates();
    };
    /**
     * @return {?}
     */
    EmailSettingsComponent.prototype.refreshTemplates = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.settings.email_addon === '@materia/mailjet') {
            this.http
                .post(this.baseUrl + "/entities/mailjet_template/queries/list", { limit: 1000 })
                .subscribe(function (data) {
                _this.templates = data.data
                    .filter(function (row) { return row.OwnerId !== 1; })
                    .map(function (row) {
                    return {
                        name: row.Name,
                        id: row.ID
                    };
                });
            });
        }
        else if (this.settings.email_addon === '@materia/sendgrid') {
            this.http
                .post(this.baseUrl + "/entities/sendgrid_template/queries/listTemplates", {})
                .subscribe(function (data) {
                _this.templates = data && data.data && data.data;
                // TODO...
            });
        }
    };
    /**
     * @return {?}
     */
    EmailSettingsComponent.prototype.saveEmailSettings = /**
     * @return {?}
     */
    function () {
        if (this.emailForm.valid) {
            this.save.emit(this.emailForm.value);
        }
    };
    /**
     * @return {?}
     */
    EmailSettingsComponent.prototype.cancel = /**
     * @return {?}
     */
    function () {
        this.hide.emit();
    };
    EmailSettingsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'materia-user-management-email-settings',
                    template: "<form [formGroup]=\"emailForm\">\n\t<mat-accordion>\n\t\t<mat-expansion-panel #signupPanel [expanded]=\"true\" fxLayout=\"column\">\n\t\t\t<mat-expansion-panel-header>\n\t\t\t\tSignup\n\t\t\t</mat-expansion-panel-header>\n\n\t\t\t<div class=\"panel-content\" fxLayout=\"column\">\n\t\t\t\t<mat-form-field>\n\t\t\t\t\t<input matInput placeholder=\"Subject\" formControlName=\"subject_signup\">\n\t\t\t\t</mat-form-field>\n\n\t\t\t\t<mat-form-field>\n\t\t\t\t\t<mat-select placeholder=\"Template\" formControlName=\"template_signup\">\n\t\t\t\t\t\t<mat-option *ngFor=\"let t of templates\" [value]=\"t.id\">{{ t.name }} ({{t.id}})</mat-option>\n\t\t\t\t\t</mat-select>\n\t\t\t\t</mat-form-field>\n\t\t\t\t<mat-form-field>\n\t\t\t\t\t<input matInput placeholder=\"Redirect URL\" formControlName=\"redirect_signup\">\n\t\t\t\t</mat-form-field>\n\t\t\t</div>\n\t\t</mat-expansion-panel>\n\n\t\t<mat-expansion-panel #changeEmailPanel>\n\t\t\t<mat-expansion-panel-header>\n\t\t\t\tChange email\n\t\t\t</mat-expansion-panel-header>\n\n\t\t\t<div class=\"panel-content\" fxLayout=\"column\">\n\t\t\t\t<mat-form-field>\n\t\t\t\t\t<input matInput placeholder=\"Subject\" formControlName=\"subject_change_email\">\n\t\t\t\t</mat-form-field>\n\n\t\t\t\t<mat-form-field>\n\t\t\t\t\t<mat-select placeholder=\"Template\" formControlName=\"template_change_email\">\n\t\t\t\t\t\t<mat-option *ngFor=\"let t of templates\" [value]=\"t.id\">{{ t.name }} ({{t.id}})</mat-option>\n\t\t\t\t\t</mat-select>\n\t\t\t\t</mat-form-field>\n\t\t\t\t<mat-form-field>\n\t\t\t\t\t<input matInput placeholder=\"Redirect URL\" formControlName=\"redirect_change_email\">\n\t\t\t\t</mat-form-field>\n\t\t\t</div>\n\t\t</mat-expansion-panel>\n\n\t\t<mat-expansion-panel #lostPasswordPanel fxLayout=\"column\">\n\t\t\t<mat-expansion-panel-header>\n\t\t\t\tLost password\n\t\t\t</mat-expansion-panel-header>\n\t\t\t<div class=\"panel-content\" fxLayout=\"column\">\n\n\t\t\t\t<mat-form-field>\n\t\t\t\t\t<input matInput placeholder=\"Subject\" formControlName=\"subject_lost_password\">\n\t\t\t\t</mat-form-field>\n\t\t\t\t<mat-form-field>\n\t\t\t\t\t<mat-select placeholder=\"Template\" formControlName=\"template_lost_password\">\n\t\t\t\t\t\t<mat-option *ngFor=\"let t of templates\" [value]=\"t.id\">{{ t.name }} ({{t.id}})</mat-option>\n\t\t\t\t\t</mat-select>\n\t\t\t\t</mat-form-field>\n\t\t\t\t<mat-form-field>\n\t\t\t\t\t<input matInput placeholder=\"Redirect URL\" formControlName=\"redirect_lost_password\">\n\t\t\t\t</mat-form-field>\n\t\t\t</div>\n\t\t</mat-expansion-panel>\n\t</mat-accordion>\n\t<div style=\"margin: 16px 0; text-align: center;\">\n\t\t<button mat-raised-button color=\"primary\" (click)=\"saveEmailSettings()\" style=\"margin-right: 8px\">Save</button>\n\t\t<button mat-raised-button (click)=\"hide.emit()\">Cancel</button>\n\t</div>\n</form>",
                    styles: [".panel-content{margin:0 24px!important}"]
                },] },
    ];
    /** @nocollapse */
    EmailSettingsComponent.ctorParameters = function () { return [
        { type: HttpClient, },
        { type: FormBuilder, },
    ]; };
    EmailSettingsComponent.propDecorators = {
        "settings": [{ type: Input },],
        "baseUrl": [{ type: Input },],
        "hide": [{ type: Output },],
        "save": [{ type: Output },],
        "signupPanel": [{ type: ViewChild, args: ['signupPanel',] },],
        "changeEmailPanel": [{ type: ViewChild, args: ['changeEmailPanel',] },],
        "lostPasswordPanel": [{ type: ViewChild, args: ['lostPasswordPanel',] },],
    };
    return EmailSettingsComponent;
}());
export { EmailSettingsComponent };
function EmailSettingsComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    EmailSettingsComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    EmailSettingsComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    EmailSettingsComponent.propDecorators;
    /** @type {?} */
    EmailSettingsComponent.prototype.settings;
    /** @type {?} */
    EmailSettingsComponent.prototype.baseUrl;
    /** @type {?} */
    EmailSettingsComponent.prototype.emailForm;
    /** @type {?} */
    EmailSettingsComponent.prototype.templates;
    /** @type {?} */
    EmailSettingsComponent.prototype.hide;
    /** @type {?} */
    EmailSettingsComponent.prototype.save;
    /** @type {?} */
    EmailSettingsComponent.prototype.signupPanel;
    /** @type {?} */
    EmailSettingsComponent.prototype.changeEmailPanel;
    /** @type {?} */
    EmailSettingsComponent.prototype.lostPasswordPanel;
    /** @type {?} */
    EmailSettingsComponent.prototype.http;
    /** @type {?} */
    EmailSettingsComponent.prototype.form;
}
