// @materia/server/lib/api/controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
const fs_extra_1 = require("fs-extra");
const error_1 = require("../error");
class Controller {
    constructor(app) {
        this.app = app;
    }
    load(basePath, controller) {
        const ctrlPath = require.resolve(path_1.join(basePath, 'server', 'controllers', controller + '.ctrl.js'));
        try {
            if (require.cache[ctrlPath]) {
                delete require.cache[ctrlPath];
            }
            this.ctrlClass = require(ctrlPath);
            this.ctrlStr = fs_extra_1.readFileSync(ctrlPath, 'utf-8').toString();
            delete this.ctrlInstance;
        }
        catch (e) {
            const err = new error_1.MateriaError(`Could not load controller ${controller}: ${e}`);
            err.originalError = e;
            throw err;
        }
    }
    instance() {
        if (!this.ctrlInstance) {
            this.ctrlInstance = new this.ctrlClass(this.app);
        }
        return this.ctrlInstance;
    }
}
exports.Controller = Controller;
