// @materia/users/client/esm2015/emails-settings/email-settings.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { FormBuilder, Validators } from '@angular/forms';
import { MatExpansionPanel } from '@angular/material';
export class EmailSettingsComponent {
    /**
     * @param {?} http
     * @param {?} form
     */
    constructor(http, form) {
        this.http = http;
        this.form = form;
        this.hide = new EventEmitter();
        this.save = new EventEmitter();
    }
    /**
     * @param {?} property
     * @param {?} defaultValue
     * @return {?}
     */
    getSettingsProperty(property, defaultValue) {
        return (this.settings && this.settings[property]) || defaultValue;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.emailForm = this.form.group({
            method: [
                this.getSettingsProperty('method', 'session'),
                Validators.required
            ],
            static_salt: [
                this.getSettingsProperty('static_salt', ''),
                Validators.required
            ],
            user_profile_enabled: [
                this.getSettingsProperty('user_profile_enabled', 'false')
            ],
            user_profile_entity: [
                this.getSettingsProperty('user_profile_entity', '')
            ],
            email_verification: [
                this.getSettingsProperty('email_verification', false)
            ],
            email_addon: [this.getSettingsProperty('email_addon', false)],
            template_signup: [this.getSettingsProperty('template_signup', '')],
            redirect_signup: [this.getSettingsProperty('redirect_signup', '')],
            subject_signup: [this.getSettingsProperty('subject_signup', '')],
            subject_lost_password: [this.getSettingsProperty('subject_lost_password', '')],
            subject_change_email: [this.getSettingsProperty('subject_change_email', '')],
            template_lost_password: [
                this.getSettingsProperty('template_lost_password', '')
            ],
            redirect_lost_password: [
                this.getSettingsProperty('redirect_lost_password', '')
            ],
            template_change_email: [
                this.getSettingsProperty('template_change_email', '')
            ],
            redirect_change_email: [
                this.getSettingsProperty('redirect_change_email', '')
            ]
        });
        this.refreshTemplates();
    }
    /**
     * @return {?}
     */
    refreshTemplates() {
        if (this.settings.email_addon === '@materia/mailjet') {
            this.http
                .post(`${this.baseUrl}/entities/mailjet_template/queries/list`, { limit: 1000 })
                .subscribe(data => {
                this.templates = data.data
                    .filter(row => row.OwnerId !== 1)
                    .map(row => {
                    return {
                        name: row.Name,
                        id: row.ID
                    };
                });
            });
        }
        else if (this.settings.email_addon === '@materia/sendgrid') {
            this.http
                .post(`${this.baseUrl}/entities/sendgrid_template/queries/listTemplates`, {})
                .subscribe(data => {
                this.templates = data && data.data && data.data;
                // TODO...
            });
        }
    }
    /**
     * @return {?}
     */
    saveEmailSettings() {
        if (this.emailForm.valid) {
            this.save.emit(this.emailForm.value);
        }
    }
    /**
     * @return {?}
     */
    cancel() {
        this.hide.emit();
    }
}
EmailSettingsComponent.decorators = [
    { type: Component, args: [{
                selector: 'materia-user-management-email-settings',
                template: `<form [formGroup]="emailForm">
	<mat-accordion>
		<mat-expansion-panel #signupPanel [expanded]="true" fxLayout="column">
			<mat-expansion-panel-header>
				Signup
			</mat-expansion-panel-header>

			<div class="panel-content" fxLayout="column">
				<mat-form-field>
					<input matInput placeholder="Subject" formControlName="subject_signup">
				</mat-form-field>

				<mat-form-field>
					<mat-select placeholder="Template" formControlName="template_signup">
						<mat-option *ngFor="let t of templates" [value]="t.id">{{ t.name }} ({{t.id}})</mat-option>
					</mat-select>
				</mat-form-field>
				<mat-form-field>
					<input matInput placeholder="Redirect URL" formControlName="redirect_signup">
				</mat-form-field>
			</div>
		</mat-expansion-panel>

		<mat-expansion-panel #changeEmailPanel>
			<mat-expansion-panel-header>
				Change email
			</mat-expansion-panel-header>

			<div class="panel-content" fxLayout="column">
				<mat-form-field>
					<input matInput placeholder="Subject" formControlName="subject_change_email">
				</mat-form-field>

				<mat-form-field>
					<mat-select placeholder="Template" formControlName="template_change_email">
						<mat-option *ngFor="let t of templates" [value]="t.id">{{ t.name }} ({{t.id}})</mat-option>
					</mat-select>
				</mat-form-field>
				<mat-form-field>
					<input matInput placeholder="Redirect URL" formControlName="redirect_change_email">
				</mat-form-field>
			</div>
		</mat-expansion-panel>

		<mat-expansion-panel #lostPasswordPanel fxLayout="column">
			<mat-expansion-panel-header>
				Lost password
			</mat-expansion-panel-header>
			<div class="panel-content" fxLayout="column">

				<mat-form-field>
					<input matInput placeholder="Subject" formControlName="subject_lost_password">
				</mat-form-field>
				<mat-form-field>
					<mat-select placeholder="Template" formControlName="template_lost_password">
						<mat-option *ngFor="let t of templates" [value]="t.id">{{ t.name }} ({{t.id}})</mat-option>
					</mat-select>
				</mat-form-field>
				<mat-form-field>
					<input matInput placeholder="Redirect URL" formControlName="redirect_lost_password">
				</mat-form-field>
			</div>
		</mat-expansion-panel>
	</mat-accordion>
	<div style="margin: 16px 0; text-align: center;">
		<button mat-raised-button color="primary" (click)="saveEmailSettings()" style="margin-right: 8px">Save</button>
		<button mat-raised-button (click)="hide.emit()">Cancel</button>
	</div>
</form>`,
                styles: [`.panel-content{margin:0 24px!important}`]
            },] },
];
/** @nocollapse */
EmailSettingsComponent.ctorParameters = () => [
    { type: HttpClient, },
    { type: FormBuilder, },
];
EmailSettingsComponent.propDecorators = {
    "settings": [{ type: Input },],
    "baseUrl": [{ type: Input },],
    "hide": [{ type: Output },],
    "save": [{ type: Output },],
    "signupPanel": [{ type: ViewChild, args: ['signupPanel',] },],
    "changeEmailPanel": [{ type: ViewChild, args: ['changeEmailPanel',] },],
    "lostPasswordPanel": [{ type: ViewChild, args: ['lostPasswordPanel',] },],
};
function EmailSettingsComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    EmailSettingsComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    EmailSettingsComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    EmailSettingsComponent.propDecorators;
    /** @type {?} */
    EmailSettingsComponent.prototype.settings;
    /** @type {?} */
    EmailSettingsComponent.prototype.baseUrl;
    /** @type {?} */
    EmailSettingsComponent.prototype.emailForm;
    /** @type {?} */
    EmailSettingsComponent.prototype.templates;
    /** @type {?} */
    EmailSettingsComponent.prototype.hide;
    /** @type {?} */
    EmailSettingsComponent.prototype.save;
    /** @type {?} */
    EmailSettingsComponent.prototype.signupPanel;
    /** @type {?} */
    EmailSettingsComponent.prototype.changeEmailPanel;
    /** @type {?} */
    EmailSettingsComponent.prototype.lostPasswordPanel;
    /** @type {?} */
    EmailSettingsComponent.prototype.http;
    /** @type {?} */
    EmailSettingsComponent.prototype.form;
}
