// @materia/mailjet/server/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class MailjetAddon {
    constructor(app, config) { }
    start() {
    }
    uninstall(app) { }
}
MailjetAddon.displayName = 'Mailjet';
MailjetAddon.logo = 'https://raw.githubusercontent.com/materiahq/materia-website-content/master/logo/addons/mailjet.jpg';
MailjetAddon.installSettings = [
    {
        name: 'apikey',
        description: 'Enter your apikey',
        type: 'string',
        required: true
    },
    {
        name: 'secret',
        description: 'Enter your secret key',
        type: 'string',
        required: true
    },
    {
        name: 'from',
        description: 'Enter the email address (sender) attached with your apikey',
        type: 'string',
        required: true,
        component: 'email'
    },
    {
        name: 'name',
        description: 'Enter the name you would like to see appear on your email',
        type: 'string',
        required: true
    }
];
exports.default = MailjetAddon;
